django/db/models/fields/files.py

```diff
@@ -91,7 +91,16 @@ def open(self, mode="rb"):
     def save(self, name, content, save=True):
         name = self.field.generate_filename(self.instance, name)
         self.name = self.storage.save(name, content, max_length=self.field.max_length)
-        setattr(self.instance, self.field.attname, self.name)
+
+        # The presumption is that IO for content has already been dealt with
+        setattr(self.instance, self.field.attname, content)
+
+        # Update the name if generate_filename or the storage backend changed
+        # the name, side-stepping the descriptor to avoid re-running logic.
+        # Part of the fix for #8307
+        if self.name != getattr(self.instance, self.field.attname).name:
+            self.instance.__dict__[self.field.attname] = self.name
```

Comment:
```suggestion
        # Update the name in case generate_filename or the storage backend 
        # changed the name. Avoids re-running the descriptor logic.
        self.instance.__dict__[self.field.attname] = self.name
```
🤔 maybe, the test still passes for me and I don't see the harm of always setting it similar to before

Reply:
I went ahead and committed the suggested change.